### PR TITLE
TY: evaluate const expr array size

### PR DESCRIPTION
**Pull request by Undin**

### CODE REVIEW

**Comment by Undin** [comment]
Do we have good way to check performance of stub creation?

**Comment by vlad20012** [comment]
Is lit expr enough? What about bin expr?

File: src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt [hunk 1]
```diff
@@ -978,6 +978,38 @@ class RsMetaItemStub(
     }
 }
 
+class RsLitExprStub(
+    parent: StubElement<*>?, elementType: IStubElementType<*, *>,
+    val integerLiteralValue: String?
+) : StubBase<RsLitExpr>(parent, elementType) {
+    object Type : RsStubElementType<RsLitExprStub, RsLitExpr>("LIT_EXPR") {
+
+        override fun shouldCreateStub(node: ASTNode): Boolean {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt]: Why `createStubIfParentIsStub(node)` isn't work here?

**Comment by Undin** [comment]
Our grammar is supposed not to generate a direct object of `RsExprImpl` type, isn't it?

My problem: we should make `RsExpr` to extend `StubBasedPsiElement` (and correspondingly
`RsExprImpl` should implement `StubBasedPsiElement` somehow) because otherwise all its child classes can't implement `StubBasedPsiElement`. 

On the one hand, `RsExprImpl` has `getElementType` method and it should return `IStubElementType` (or it fails at runtime while call) and if there are real direct objects of `RsExprImpl` we must generate implementation of `IStubElementType` for `RsExpr` (test fails because there isn't such implementation). On the other hand, if we add corresponding code into `org.rust.lang.core.stubs.StubImplementationsKt.factory`, stubs for expr PSI elements will be created incorrectly.
 
So direct objects of `RsExprImpl` should not exist and there is some mistake in grammar (test fails at call of `RsExprImpl.getElementType`) or I have no idea how to create stubs for expressions.

**Comment by Undin** [comment]
@vlad20012 @matklad any tips?

**Comment by matklad** [comment]
Yeah, I think that ideally `RsExprImpl` should be an `abstract` class. However it's also true that inheritance, Grammar Kit, and stubs together don't mesh up really well.

The trick we used for TypeReferences (when they used `extends` in the grammar) was to create a single stub type for all various classes.

**Comment by Undin** [comment]
My current implementation uses single stub type for all expressions, but current grammar allows to generate direct objects of `RsExprImpl` somehow and current tests fail because of it. It is bug, isn't it?

**Comment by Undin** [comment]
Hm, no, I use different stub class, sorry
But it seems it's not a problem

**Comment by Undin** [comment]
Hmm, no, I use different stub classes, sorry
But it seems that it isn't problem in this case

**Comment by matklad** [comment]
>but current grammar generates direct objects of RsExprImpl somehow

Yeah, I *think* this should not happen (this class should be abstract), but I am not sure what's causes this.

**Comment by Undin** [comment]
Ok, thanks:)
I'll try to find out how it happens

**Comment by Undin** [comment]
Hm, looks like that
```rust
#[attr]
{}
```
in 
```rust
fn main() {
    #[attr]
    {}
}
```
generates direct object of `RsExprImpl`

**Comment by Undin** [comment]
@matklad thanks for #2039! `RsCompilerSourcesPerformance` test works!

**Comment by Undin** [comment]
Finally, it seems it works! 🎉

File: src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt [hunk 2]
```diff
@@ -534,27 +535,27 @@ private class RsFnInferenceContext(
 
     private fun tryCoerce(inferred: Ty, expected: Ty): Boolean {
         return when {
-        // Coerce array to slice
+            // Coerce array to slice
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt]: Hmm, intellij auto-format thinks differently about this

File: src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt [hunk 3]
```diff
@@ -1255,7 +1260,7 @@ private class RsFnInferenceContext(
                 ?: expr.initializer?.inferType()
                 ?: return TySlice(TyUnknown)
             expr.sizeExpr?.inferType(TyInteger(TyInteger.Kind.usize))
-            val size = calculateArraySize(expr.sizeExpr)
+            val size = calculateArraySize(expr.sizeExpr) { resolvePathExpr(it).singleOrNull()?.element }
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt]: Here we don't need any additional `resolvePathExpr`. We can just lookup already resolved path:
```rust
calculateArraySize(expr.sizeExpr) { ctx.resolvedPaths[it] }
```

File: src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt (hunk 2 above)
Comment (Undin) [reply src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt]: It's weird but I've reverted this change

**Comment by vlad20012** [comment]
I was very worried about the assumption that extra stubs (e.g. full function bodies) may be created as a side-effect of thid PR. So I write some tests that check stub existence. [Try it](https://github.com/vlad20012/intellij-rust/commit/9bf0119e29ce072c0a6ed271d2ffb0f4fa927dcf) (don't merge, just copy-paste the code). I've tried it and it seems like even expressions inside statements are stubbed =(
Or maybe my tests are broken.
c.c. @matklad

File: src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt [hunk 4]
```diff
@@ -1038,14 +1028,96 @@ class RsMetaItemStub(
...
+}
+
+class RsLitExprStub(
+    parent: StubElement<*>?, elementType: IStubElementType<*, *>,
+    val type: RsStubLiteralType?,
+    val integerLiteralValue: String?
+) : RsPlaceholderStub(parent, elementType) {
+    object Type : RsStubElementType<RsLitExprStub, RsLitExpr>("LIT_EXPR") {
+
+        override fun serialize(stub: RsLitExprStub, dataStream: StubOutputStream) {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt]: :thinking: 
```kotlin
 override fun shouldCreateStub(node: ASTNode): Boolean = createStubIfParentIsStub(node)
```

File: src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt [hunk 5]
```diff
@@ -1038,14 +1028,96 @@ class RsMetaItemStub(
...
+    }
+}
+
+class RsLitExprStub(
+    parent: StubElement<*>?, elementType: IStubElementType<*, *>,
+    val type: RsStubLiteralType?,
+    val integerLiteralValue: String?
+) : RsPlaceholderStub(parent, elementType) {
+    object Type : RsStubElementType<RsLitExprStub, RsLitExpr>("LIT_EXPR") {
+
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt]: :thinking: 
```kotlin
 override fun shouldCreateStub(node: ASTNode): Boolean = createStubIfParentIsStub(node)
```

Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt]: :thinking: 
```kotlin
override fun shouldCreateStub(node: ASTNode): Boolean = createStubIfParentIsStub(node)
```

Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt]: Yep, it fixes most cases except function tail expr

Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt]: :thinking: :thinking: 
```kotlin
override fun shouldCreateStub(node: ASTNode): Boolean = createStubIfParentIsStub(node) &&
            node.psi.parent.parent !is RsFunction
```

Comment (Undin) [reply src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt]: 🤦
I was completely sure that `RsLitExprStub.Type` is inheritor of `RsPlaceholderStub.Type` and it's already used `createStubIfParentIsStub`
Sorry, my fault

**Comment by vlad20012** [comment]
```rust
        fn foo() -> i32 { 2 + 2 }
                          //^
```
This test is still failing

**Comment by vlad20012** [comment]
Anyway, I think I should finally merge it. 4 month, wow :)
I'll add stub tests and fix tail exprs right after the merge.
bors r+

**Pull request merged**